### PR TITLE
DOC: Improve "inplace" param in Dataframe.rename() docs

### LINKED ISSUE #23076: Unclear documentation of inplace parameter for pandas.DataFrame.rename.
#### Problem description

The documentation for pandas.DataFrame.rename currently reads

> inplace : boolean, default False
> 
> Whether to return a new DataFrame. If True then value of copy is ignored.

The sentence "Whether to return a new DataFrame" suggests that setting this parameter to True means a new DataFrame will be returned. The parameter name "inplace" suggests that setting this parameter to True means a new DataFrame will **not** be returned, which is actually the case.

The only thing in the documentation which concretely specifies the true behavior is the parameter name.

A possible alternate phrasing which suggests the correct thing would be to replace the sentence "Whether to return a new DataFrame" with something like "Modify the DataFrame, instead of returning a new DataFrame."

Same thing for the base class NDFrame and its other derived classes like Panel.

### PR DESCRIPTION
**Pull request by nmusolino**
- [x] closes #23076
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by nmusolino** [comment]
As described in #23076, the description of "inplace" in the  [current documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename.html) is "backwards".

Arguably, there should be a wider cleanup to use a consistent phrasing for "inplace", but the scope of this PR is just fixing the "backwards" occurrence in this method.

Here is some "prior art" for how `inlace` is documented in DataFrame methods with a similar purpose:

[DataFrame.drop()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
```
    inplace : bool, default False
        If True, do operation inplace and return None.
```

[DataFrame.reset_index()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.reset_index.html)
```
    inplace : boolean, default False
        Modify the DataFrame in place (do not create a new object)
```

[DataFrame.rename_axis()](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename_axis.html#pandas.DataFrame.rename_axis).  This is probably from a shared NDFrame docstring.
```
    inplace : boolean, default False
         Modifies the object directly, instead of creating a new Series or DataFrame.
```

**Comment by gfyoung** [comment]
> Arguably, there should be a wider cleanup to use a consistent phrasing for inplace across pandas

That sounds like a plan.  Maybe we could do it in this PR even.

**Comment by nmusolino** [comment]
I'm willing to work on that in a future PR.  My suggestion is to get this PR in right now, in order to fix the mistaken docstring and close the associated issue.

The problem with the current `DataFrame.rename()` docstring, just to be explicit, is this:  in general, when a parameter <X> has a description "whether to <blank>", the implication is that when the parameter X is true, the implementation will "<blank>".  The current docstring does not follow that convention.

As for the wider change, I think that will take more discussion:
a.  How should the parameter description be phrased?
b.  Should the "Returns" section of each method include a note about returning None when inplace is True (where applicable)?
c.  Should the parameter description mention that internal data structures (e.g. block manager or index) might be copied even when inplace is True?

Or even, should that effort be made in light of #16529: [API/DEPR: Deprecate inplace parameter](https://github.com/pandas-dev/pandas/issues/16529)?  I just saw that deprecation proposal now.

**Comment by jreback** [comment]
cc @datapythonista

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3909,8 +3909,8 @@ def rename(self, *args, **kwargs):
         copy : boolean, default True
             Also copy underlying data
         inplace : boolean, default False
-            Whether to return a new DataFrame. If True then value of copy is
-            ignored.
+            If True, perform the operation in-place, modifying this DataFrame,
+            and return None.
```
Comment (datapythonista) [inline pandas/core/frame.py]: I don't think this is technically correct. The name is misleading, but using `inplace` doesn't necessarily mean that we're modifying the `DataFrame`, as we can be returning a copy.

I think the previous description is more accurate, except that it sounds like `inplace=True` will return a new `DataFrame`.

**Review by datapythonista — CHANGES_REQUESTED** [review]
+1 on standardizing the descriptions for `inplace`. But the proposed description is no exactly correct.

File: pandas/core/frame.py (hunk 1 above)
Comment (datapythonista) [reply pandas/core/frame.py]: If you have a `DataFrame` named `df`, using the memory in the address 1000-2000, and you do `df.operation(inplace=True)`, `df` will be modified, but the data can be now in the memory address 3000-4000.

So, technically speaking, the operation is not in place, as the result has been copied to a different location (even if the reference `df` is reassigned to the new memory).

That's why I don't like your description, as for a user who doesn't care about the memory is perfectly fine and correct. But if you think about the memory, you description is not really correct.

Comment (jreback) [reply pandas/core/frame.py]: @nmusolino there are about 2 inplace operations that *actually* operate inplace. all others, simply do an internal copy, do the op, then assign back to the top level data. inplace has never been a real operation in pandas, nor will it ever be. This is why we are deprecating it. You can certainly give a longer explanation, but discouraging usage would be good here.

Comment (datapythonista) [reply pandas/core/frame.py]: @jreback are we ready to deprecate all `inplace` parameters? A quick grep shows around 60 methods with an `inplace` parameter, so it's not so much work to start raising warnings (and update the descriptions).

Or do we prefer to wait?

Comment (gfyoung) [reply pandas/core/frame.py]: @datapythonista :

See #16529 - I think we were planning to wait until 1.0, though can always push up.

Comment (datapythonista) [reply pandas/core/frame.py]: Thanks for pointing out the issue @gfyoung, I'll continue the conversation there.

Comment (nmusolino) [reply pandas/core/frame.py]: @datapythonista, @jreback , I understand that these operations may involve copies.  I  wrote earlier that "internal data structures (e.g. block manager or index) might be copied even when inplace is True."  

I think the distinction you are trying to draw is that a `rename` or `fillna` operation does not necessarily overwrite individual elements in place; is that your point?  As a simple example, I seem to recall that Index objects are generally immutable, so any operation that changes the index would necessarily replace the index.

To be clear, however, the methods are modifying the dataframe.

How about this description?
```
    inplace : boolean, default False
        If true, rename index labels in this DataFrame and return None, 
        without creating a new DataFrame.
```

Or a simpler, more generic version:
```
    inplace : boolean, default False
        If true, modify this DataFrame and return None; does not create a 
        new DataFrame.
```

**Comment by datapythonista** [comment]
Closing in favor of #24063, where the `inplace` argument will be deprecated, and all descriptions updated.

**Pull request closed without merging**